### PR TITLE
[chore] Use check-latest in actions/setup-go

### PR DESCRIPTION
**Pull request by pellared**
Avoid problems like this one https://github.com/open-telemetry/opentelemetry-go/pull/4607 by using [`check-latest-version`](https://github.com/actions/setup-go#check-latest-version)

Opportunistic improvements:
- use [`cache-dependency-path`](https://github.com/actions/setup-go#caching-dependency-files-and-build-outputs) when possible
- remove " Setup Environment" steps as they are [no longer needed since v3](https://github.com/actions/setup-go#v3)

### CODE REVIEW

**Comment by pellared** [comment]
When `check-latest: true` is used we can change the `go-version` values to `"1.20"` and `"1.21"`. Can we do this (upvote/downvote)?

**Comment by pellared** [comment]
> When `check-latest: true` is used we can change the `go-version` values to `"1.20"` and `"1.21"`. Can I do this (upvote) or do you prefer to still keep the explicit minimal patch version (downvote)?

Done https://github.com/open-telemetry/opentelemetry-go/pull/4617/commits/aa526b7b575b0bed9da52a4e70dd9d6afe4016e9

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -14,29 +14,19 @@ env:
...
           go-version: ${{ env.DEFAULT_GO_VERSION }}
+          check-latest: true
           cache-dependency-path: "**/go.sum"
-      - name: Tools cache
-        uses: actions/cache@v3
-        env:
-          cache-name: go-tools-cache
-        with:
-          path: ~/.tools
-          key: ${{ runner.os }}-${{ env.cache-name }}-${{ hashFiles('./internal/tools/**') }}
```
Comment (MadVikingGod) [inline .github/workflows/ci.yml:39]: Do you have any insight on why this was here and why it's ok to remove it?

**Review by MadVikingGod — APPROVED** [review]

File: .github/workflows/ci.yml (hunk 1 above)
Comment (pellared) [reply .github/workflows/ci.yml:39]: I have not spotted that the `path: ~/.tools` I will revert it.

Comment (pellared) [reply .github/workflows/ci.yml]: https://github.com/open-telemetry/opentelemetry-go/pull/4617/commits/4cb1bbf8701a933fa90ef92baa30eb1bee66384d

**Comment by pellared** [comment]
Now the `lint` job is ~1 min faster 💪 

[main](https://github.com/open-telemetry/opentelemetry-go/actions/runs/6511461650/job/17687207918) 5m14s
[PR](https://github.com/open-telemetry/opentelemetry-go/actions/runs/6511576277/job/17687723993?pr=4617) 3m54s

**Comment by MrAlias** [comment]
> When `check-latest: true` is used we can change the `go-version` values to `"1.20"` and `"1.21"`. Can I do this (upvote) or do you prefer to still keep the explicit minimal patch version (downvote)?

I would rather keep the minimums pegged above versions with CVEs. This will help ensure no issues are re-introduced.

**Comment by pellared** [comment]
> > When `check-latest: true` is used we can change the `go-version` values to `"1.20"` and `"1.21"`. Can I do this (upvote) or do you prefer to still keep the explicit minimal patch version (downvote)?
> 
> I would rather keep the minimums pegged above versions with CVEs. This will help ensure no issues are re-introduced.

What do you mean by re-introducing? A regression bug that we remove `check-latest: true`? If this would happen we would simply add it back. Do we plan bump the version each time a Go patch is released?

**Comment by pellared** [comment]
> > When `check-latest: true` is used we can change the `go-version` values to `"1.20"` and `"1.21"`. Can I do this (upvote) or do you prefer to still keep the explicit minimal patch version (downvote)?
> 
> I would rather keep the minimums pegged above versions with CVEs. This will help ensure no issues are re-introduced.

@MrAlias, what do you mean by re-introducing? A regression bug that we remove `check-latest: true`? If this would happen we would simply add it back.

Is your intention that we bump the version each time a Go patch is released? If so then why maybe we should be pinning to the Go version instead e.g. `go-version: "1.21.3"? This would give the advantage of having more reproducible builds.

**Review by hanyuancheung — APPROVED** [review]

**Comment by pellared** [comment]
Merging as at least CI is quicker. I created https://github.com/open-telemetry/opentelemetry-go/issues/4639 for potentially addressing https://github.com/open-telemetry/opentelemetry-go/pull/4617#issuecomment-1762162711.

**Comment by pellared** [comment]
> I would have appreciated it if all feedback was addressed prior to this merging, in accordance with our [policy](https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md#how-to-get-prs-merged).

1. It was not clear that it isa blocker feedback (there was no "request changes").
2. There was no consensus as @MadVikingGod upvoted my proposal.

I found it more beneficial to merge as it is and created an issue as part of addressing the feedback. I had no bad intentions 😉

**Pull request merged**